### PR TITLE
Implement SE-0062: Referencing Objective-C key-paths

### PR DESCRIPTION
**Pull request by DougGregor**
#### What's in this pull request?

This pull request implements [SE-0062: Referencing Objective-C key-paths](https://github.com/apple/swift-evolution/blob/master/proposals/0062-objc-keypaths.md)
#### Resolved bug number: ([SR-1237](https://bugs.swift.org/browse/SR-1237))

Also: rdar://problem/25710611

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

 **Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| OS X platform | @swift-ci Please test OS X platform |
| Linux platform | @swift-ci Please test Linux platform |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please test

File: include/swift/AST/Expr.h [hunk 1]
```diff
@@ -311,6 +311,19 @@ class alignas(8) Expr {
...
+    friend class ObjCKeyPathExpr;
+    unsigned : NumExprBits;
+
+    /// The number of components in the selector path.
+    unsigned NumComponents : 8;
+
+    /// Whether the names have corresponding source locations.
+    unsigned HaveSourceLocations : 1;
+  };
+  enum { NumObjCKeyPathExprBits = NumExprBits + 17 };
```
Comment (rudkx) [inline include/swift/AST/Expr.h]: Where does 17 come from?

Comment (DougGregor) [inline include/swift/AST/Expr.h]: Good catch! I removed a field and didn't update the bit count.

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by jopamer** [comment]
Looks great!

**Comment by DougGregor** [comment]
Added ChangeLog entry, now we're good to go.

**Pull request merged**